src/ros_robot_localization_listener.cpp

```diff
@@ -0,0 +1,442 @@
+  EstimatorState state;
+
+  // Set its time stamp and the state received from the messages
+  state.time_stamp = odom.header.stamp.toSec();
+
+  // Get the base frame id from the odom message
+  if ( base_frame_id_.empty() )
+    base_frame_id_ = odom.child_frame_id;
+
+  // Get the world frame id from the odom message
+  if ( world_frame_id_.empty() )
+    world_frame_id_ = odom.header.frame_id;
+
+  // Pose: Position
+  state.state(StateMemberX) = odom.pose.pose.position.x;
+  state.state(StateMemberY) = odom.pose.pose.position.y;
+  state.state(StateMemberZ) = odom.pose.pose.position.z;
+
+  // Pose: Orientation
+  tf2::Quaternion orientation_quat;
+  tf2::fromMsg(odom.pose.pose.orientation, orientation_quat);
+  double roll, pitch, yaw;
+  RosFilterUtilities::quatToRPY(orientation_quat, roll, pitch, yaw);
+
+  state.state(StateMemberRoll) = roll;
+  state.state(StateMemberPitch) = pitch;
+  state.state(StateMemberYaw) = yaw;
+
+  // Pose: Covariance
+  for ( unsigned int i = 0; i < POSE_SIZE; i++ )
+  {
+    for ( unsigned int j = 0; j < POSE_SIZE; j++ )
+    {
+      state.covariance(POSITION_OFFSET + i, POSITION_OFFSET + j) = odom.pose.covariance[i*POSE_SIZE + j];
+    }
+  }
+
+  // Velocity: Linear
+  state.state(StateMemberVx) = odom.twist.twist.linear.x;
+  state.state(StateMemberVy) = odom.twist.twist.linear.y;
+  state.state(StateMemberVz) = odom.twist.twist.linear.z;
+
+  // Velocity: Angular
+  state.state(StateMemberVroll) = odom.twist.twist.angular.x;
+  state.state(StateMemberVpitch) = odom.twist.twist.angular.y;
+  state.state(StateMemberVyaw) = odom.twist.twist.angular.z;
+
+  // Velocity: Covariance
+  for ( unsigned int i = 0; i < TWIST_SIZE; i++ )
+  {
+    for ( unsigned int j = 0; j < TWIST_SIZE; j++ )
+    {
+      state.covariance(POSITION_V_OFFSET + i, POSITION_V_OFFSET + j) = odom.twist.covariance[i*TWIST_SIZE + j];
+    }
+  }
+
+  // Acceleration: Linear
+  state.state(StateMemberAx) = accel.accel.accel.linear.x;
+  state.state(StateMemberAy) = accel.accel.accel.linear.y;
+  state.state(StateMemberAz) = accel.accel.accel.linear.z;
+
+  // Acceleration: Angular is not available in state
+
+  // Acceleration: Covariance
+  for ( unsigned int i = 0; i < ACCELERATION_SIZE; i++ )
+  {
+    for ( unsigned int j = 0; j < ACCELERATION_SIZE; j++ )
+    {
+      state.covariance(POSITION_A_OFFSET + i, POSITION_A_OFFSET + j) = accel.accel.covariance[i*TWIST_SIZE + j];
+    }
+  }
+
+  // Add the state to the buffer, so that we can later interpolate between this and earlier states
+  estimator_->setState(state);
+
+  return;
+}
+
+bool RosRobotLocalizationListener::getState(const double time,
+                                            const std::string& frame_id,
+                                            Eigen::VectorXd& state,
+                                            Eigen::MatrixXd& covariance,
+                                            std::string world_frame_id)
+{
+  EstimatorState estimator_state;
```

Comment:
This is going to need some check for the nature of the requested transform. Right now, requesting the state of "world" would give a reasonable result without complaining, but it would use the robot state to transform between odom and base, and then tf to transform between base and utm. We can either choose to throw an error if base is not a parent of the requested frame id, or only transform using tf if this is the case.